### PR TITLE
[BEAM-3900] Support for all windowing options.

### PR DESCRIPTION
**Pull request by dmvk**
**Please** add a meaningful description for your change here

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by mareksimunek** [comment]
What about make from `WindowedOutputBuilder` abstract class so we wont repeat same methods from `WindowedOutput` interface in every operator.

Will there be any cons to that?

**Comment by dmvk** [comment]
@mareksimunek this was not possible due to java generic limitations. I've introduced WindowState class, that should minimize the pain of maintaining window builder logic.

**Review by mareksimunek — APPROVED** [review]
LGTM

File: sdks/java/extensions/euphoria/euphoria-core/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/client/operator/WindowState.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+   *
+   * @return maybe window
+   */
+  Optional<Window<T>> getWindow() {
+    return Optional.ofNullable(window);
+  }
+
+  /**
+   * Set underlying window. This is useful for building composite operators, such as {@link
+   * ReduceByKey}.
```
Comment (VaclavPlajt) [inline sdks/java/extensions/euphoria/euphoria-core/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/client/operator/WindowState.java]: `RBK` is not composite operator at lest from `CompositeOperator` point of view.

File: sdks/java/extensions/euphoria/euphoria-core/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/client/operator/WindowState.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import org.apache.beam.sdk.transforms.windowing.WindowFn;
+import org.apache.beam.sdk.values.WindowingStrategy;
+import org.joda.time.Duration;
+
+/**
+ * Holds state of {@link Window} builder.
+ *
+ * @param <T> type of windowed element
+ */
+class WindowState<T>
```
Comment (VaclavPlajt) [inline sdks/java/extensions/euphoria/euphoria-core/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/client/operator/WindowState.java]: Why not to call it `WindowBuilder` when it is in fact shared window builder.

**Review by VaclavPlajt — APPROVED** [review]

**Pull request merged**